### PR TITLE
Make resolve_object not require auth #3685

### PR DESCRIPTION
**Pull request by marsara9**
Resolves issue #3685

If user isn't authenticated with resolve_object, only allow a local search instead of possibly making an http request.

Changing resolve_object to take an optional auth parameter.  When auth is null, the request is only allowed to search the local database rather than do a remote lookup.

This is primarily to resolve a few issue:
1) In my own project https://www.github.com/marsara9/lemmy-search using `/api/v3/post...` on every Lemmy instance in the fediverse, is extremely slow.  Current estimate, is around ~13 years worst case, if no new posts get added.
2) This will be first step in enabling lemmy-ui to be able to rewrite links to post to always be in the user's home instance https://github.com/LemmyNet/lemmy/issues/2987.

Original PR for reference: https://github.com/LemmyNet/lemmy/pull/3706 (as it was easier to decline that one and open a new one with the requested changes).

### CODE REVIEW

**Comment by marsara9** [comment]
It doesn't look like this is actually blocking invalid tokens.  I'll push an update tonight to fix that.

**Comment by marsara9** [comment]
Fixed the issue where I wasn't properly locking invalid tokens.

File: crates/apub/src/fetcher/search.rs [hunk 1]
```diff
@@ -44,6 +44,18 @@ pub(crate) async fn search_query_to_object_id(
...
+/// Converts a search query to an object id.  The query MUST bbe a URL which will bbe treated
+/// as the ObjectId directly.  If the query is a webfinger identifier (@user@example.com or
+/// !community@example.com) this method will return an error.
+#[tracing::instrument(skip_all)]
+pub(crate) async fn search_query_to_object_id_local(
+  query: &str,
+  context: &Data<LemmyContext>,
+) -> Result<SearchableObjects, LemmyError> {
+  let url = Url::parse(query)?;
+  ObjectId::from(url).dereference_local(context).await
```
Comment (Nutomic) [inline crates/apub/src/fetcher/search.rs:56]: You can directly call `ObjectId::parse` which calls `Url::parse` internally. And in fact it doesnt make much sense to have a separate function for this single line.

**Review by Nutomic — APPROVED** [review]

File: crates/api_common/src/site.rs [hunk 2]
```diff
@@ -84,7 +84,7 @@ pub struct SearchResponse {
 pub struct ResolveObject {
   /// Can be the full url, or a shortened version like: !fediverse@lemmy.ml
   pub q: String,
-  pub auth: Sensitive<String>,
+  pub auth: Option<Sensitive<String>>,
```
Comment (dessalines) [inline crates/api_common/src/site.rs:87]: API change, but it shouldn't be breaking.

**Review by dessalines — APPROVED** [review]

**Pull request merged**